lib/rules/no-undefined.js

```diff
@@ -21,15 +21,76 @@ module.exports = {
 
     create(context) {
 
+        /**
+         * Checks whether the given reference is 'undefined' or not.
+         *
+         * @param {escope.Reference} reference - The reference to check.
+         * @returns {boolean} `true` if the reference is 'undefined'.
+         */
+        function isUndefinedReference(reference) {
+            const id = reference.identifier;
+
+            return id && id.name === "undefined";
+        }
+
+        /**
+         * Report an invalid "undefined" identifier node.
+         * @param {ASTNode} node The node to report.
+         * @returns {void}
+         */
+        function report(node) {
+            context.report({
+                node,
+                message: "Unexpected use of undefined."
+            });
+        }
+
+        /**
+         * Checks the given scope for references to `undefined` and reports
+         * all references found.
+         * @param {escope.Scope} scope The scope to check.
+         * @returns {void}
+         */
+        function checkScope(scope) {
+            const undefinedVar = scope.variables
+                ? scope.variables.filter(variable => variable.name === "undefined")[0]
+                : null;
+
+            const references = undefinedVar
+                ? undefinedVar.references
+                : [];
+
+            const defs = undefinedVar
+                ? undefinedVar.defs
+                : [];
+
+            // Report non-initializing references (those are covered in defs below)
+            references
+                .filter(ref => !ref.init)
+                .forEach(ref => report(ref.identifier));
+
+            defs.forEach(def => report(def.name));
+        }
+
         return {
+            "Program:exit"() {
+                const globalScope = context.getScope();
+
+                /* 'scope.through' includes all references to undefined
+                 * variables. If the variable 'undefined' is not defined, it
+                 * can be found in 'scope.through'.
+                 */
+                globalScope.through
+                    .filter(isUndefinedReference)
+                    .forEach(ref => report(ref.identifier));
```

Comment:
@mysticatea You're right. It seems that `scope.through` is for when you want to catch variables that were assigned in a lower scope but which have percolated into the current scope without having been originally defined there. Since I'm iterating over all scopes anyway, the variable will be iterated over once I get to its immediate containing scope. I've just removed this logic and added a test case that was missing, and all is well.